### ISSUE TITLE
Empty batches for sparse datasets

### ISSUE THREAD

**Issue by DomInvivo** [datamol-io/graphium#413]
I have re-thought the issue related to having some tasks with empty labels. And after thinking about it, it seems that it's not a problem at all for the training, only for the logging of the curves. Here are my thoughts.

1. When doing gradient accumulations, it's not the losses that are averaged, it's the gradients
2. When a given loss is 0, it means the model is perfect, and the gradient is therefore 0
3. There will never be an occurrence where all tasks have no labels, because every molecule comes with at least 1 label on 1 task
4. Therefore, when computing the gradient, only the task-head with no labels has no gradient. All other modules still have gradients
5. When averaging the gradient contribution of all accumulated-batches, it will be exactly the same as if we had a much larger batch, but multiplied by a factor num_non_empty_batches / num_batches . 
6. This would only mean that the task with empty labels will contribute less to the main GNN. But since it's the only gradient going to its own head, then it's fine.
7. With enough number of accumulations, and with using a smaller value of epoch_sampling_fraction, we can mitigate this problem
8. We can also avoid the problem by using reduction="sum" in the loss initialisation, but this will make imbalance even worse, unless we re-reduce it later on
9. Only the logging of the losses can be problematic: We can filter out the zero-losses per-task in on_train_batch_end, as proposed by @zhiyil-graphcore  (this ONLY affects logging, the backward is already computed)

In summary, it will simply mean that a more sparse task will contribute less to the global GNN gradient, which is the case anyways if we had a very large batch instead of many small batches. But the task head will still train very well. So I think we can simply ignore the issue, use small epoch_sampling_fraction for the large datasets, and just fix the average loss used by the logging.

If you have comments or agree/disagree, happy to hear your thoughts

**Comment by DomInvivo** [comment]
After discussing this, we found that the training dynamics is good, and the task-specific loss filtering was applied in PR #391 . This issue will be closed
